Realm/Realm/Handles/SessionHandle.cs

```diff
@@ -225,14 +253,82 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
                     exception = new SessionException(messageString, errorCode);
                 }
 
-                Session.RaiseError(session, exception);
+                if (syncConfigurationBase.OnSessionError != null)
+                {
+                    syncConfigurationBase.OnSessionError.Invoke(session, exception);
+                }
+                else
+                {
+                    // after deprecation: this will need to go when Session.Error is fully deprecated
+                    Session.RaiseError(session, exception);
+                }
             }
             catch (Exception ex)
             {
                 Logger.Default.Log(LogLevel.Warn, $"An error has occurred while handling a session error: {ex}");
             }
         }
 
+        [MonoPInvokeCallback(typeof(NativeMethods.NotifyBeforeClientReset))]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "CA2000:Dispose objects before losing scope", Justification = "Unowned Realm don't need to be disposed")]
+        private static bool NotifyBeforeClientReset(IntPtr beforeFrozen, IntPtr managedSyncConfigurationHandle)
+        {
+            var syncConfigHandle = GCHandle.FromIntPtr(managedSyncConfigurationHandle);
+            var syncConfiguration = (SyncConfigurationBase)syncConfigHandle.Target;
+
+            if (syncConfiguration.ClientResetHandler == null)
+            {
+                return true;
+            }
+
+            var discardLocalResetHandler = (DiscardLocalResetHandler)syncConfiguration.ClientResetHandler;
+            var schema = syncConfiguration.GetSchema();
+            var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfiguration, schema);
```

Comment:
Why do we construct the realm before we even verify that `OnBeforeReset` is not null?